os/net/nbr-table.c

```diff
@@ -179,101 +168,148 @@ nbr_set_bit(uint8_t *bitmap, nbr_table_t *table, nbr_table_item_t *item, int val
-        if(!locked) {
-          int used = used_map[item_index];
-          int used_count = 0;
-          /* Count how many tables are using this item */
-          while(used != 0) {
-            if((used & 1) == 1) {
-              used_count++;
-            }
-          used >>= 1;
-          }
-          /* Find least used item */
-          if(least_used_key == NULL || used_count < least_used_count) {
-            least_used_key = key;
-            least_used_count = used_count;
-            if(used_count == 0) { /* We won't find any least used item */
-              break;
-            }
-          }
-        }
-        key = list_item_next(key);
+  /* No more space, try to free a neighbor */
+  for(k = list_head(nbr_table_keys); k != NULL; k = list_item_next(k)) {
+    int item_index = index_from_key(k);
+    int locked = locked_map[item_index];
+    /* Never delete a locked item */
+    if(!locked) {
+      if(worst_lladdr == NULL) {
+        worst_lladdr = &k->lladdr;
+      } else {
+        worst_lladdr = NBR_TABLE_GC_GET_WORST(worst_lladdr, &k->lladdr);
       }
     }
+  }
 
-    if(least_used_key == NULL) {
-      /* We haven't found any unlocked item, allocation fails */
+  /* Finally compare against current candidate for insertion */
+  if(worst_lladdr != NULL && NBR_TABLE_CAN_ACCEPT_NEW(new, worst_lladdr, reason, data)) {
+    return worst_lladdr;
+  } else {
+    return NULL;
+  }
+}
+/*---------------------------------------------------------------------------*/
+static bool
+entry_is_allowed(nbr_table_t *table, const linkaddr_t *lladdr,
+                 nbr_table_reason_t reason, void *data,
+                 const linkaddr_t **to_be_removed_ptr)
+{
+  bool ret;
+  const linkaddr_t *to_be_removed = NULL;
+
+  if(nbr_table_get_from_lladdr(table, lladdr) != NULL) {
+    /* Already present in the given table */
+    ret = true;
+  } else {
+    if(index_from_lladdr(lladdr) != -1
+       || memb_numfree(&neighbor_addr_mem) > 0) {
+      /* lladdr already present globally, or there is space for a new lladdr,
+       * check if entry can be added */
+      ret = NBR_TABLE_CAN_ACCEPT_NEW(lladdr, NULL, reason, data);
+    } else {
+      ret = (to_be_removed = select_for_removal(lladdr, reason, data)) != NULL;
+    }
+  }
+  if(to_be_removed_ptr != NULL) {
+    *to_be_removed_ptr = to_be_removed;
+  }
+  return ret;
+}
+/*---------------------------------------------------------------------------*/
+bool
+nbr_table_entry_is_allowed(nbr_table_t *table, const linkaddr_t *lladdr,
+                           nbr_table_reason_t reason, void *data)
+{
+  return entry_is_allowed(table, lladdr, reason, data, NULL);
+}
+/*---------------------------------------------------------------------------*/
+static nbr_table_key_t *
+nbr_table_allocate(nbr_table_reason_t reason, void *data, const linkaddr_t *to_be_removed_lladdr)
+{
+  if(memb_numfree(&neighbor_addr_mem) > 0) {
+    return memb_alloc(&neighbor_addr_mem);
```

Comment:
Why the change to search for free memory twice instead of simply calling `memb_alloc()` and see if it returns `NULL` or not?

Reply:
Yes good point, done now.